File: cpp/817_链表组件_medium.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def numComponents(self, head, G):
        set_G = set(G)
        p = head
        ans = 0
        while p.next:
            if p.val in set_G and p.next.val not in set_G:
                # print(p.val, p.next.val)
                ans += 1
            p = p.next
        if p.val in set_G:
            ans += 1
        return ans

File: cpp/817_链表组件_medium.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    int numComponents(ListNode* head, vector<int>& G) {
        unordered_map<int, int> mapG;
        for(int g:G) mapG[g] = 1;
        ListNode *p = head;
        int ans = 0;
        while(p->next){
            if(mapG.find(p->val) != mapG.end() && mapG.find(p->next->val) == mapG.end())
                ans++;
            p = p->next;
        }
        if(mapG.find(p->val) != mapG.end())
            ans++;
        return ans;
    }
};